// windicss/plugin/typography/index.js
'use strict';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    if (typeof b !== "function" && b !== null)
        throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __spreadArray(to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
}

function uniq(array) {
    return Array.from(new Set(array));
}
function castArray(value) {
    if (!arguments.length) {
        return [];
    }
    return Array.isArray(value) ? value : [value];
}
function isUsableColor(color, values) {
    return Boolean(values && typeof values === 'object' && color !== 'gray' && values[600]);
}

var createPlugin = function (plugin, config) {
    return {
        handler: plugin,
        config: config,
    };
};
createPlugin.withOptions = function (pluginFunction, configFunction) {
    if (configFunction === void 0) { configFunction = function () { return ({}); }; }
    var optionsFunction = function (options) {
        if (options === void 0) { options = {}; }
        return {
            __options: options,
            handler: pluginFunction(options),
            config: configFunction(options),
        };
    };
    optionsFunction.__isOptionsFunction = true;
    // Expose plugin dependencies so that `object-hash` returns a different
    // value if anything here changes, to ensure a rebuild is triggered.
    optionsFunction.__pluginFunction = pluginFunction;
    optionsFunction.__configFunction = configFunction;
    return optionsFunction;
};

var colors = {
    black: '#000',
    white: '#fff',
    rose: {
        50: '#fff1f2',
        100: '#ffe4e6',
        200: '#fecdd3',
        300: '#fda4af',
        400: '#fb7185',
        500: '#f43f5e',
        600: '#e11d48',
        700: '#be123c',
        800: '#9f1239',
        900: '#881337',
    },
    pink: {
        50: '#fdf2f8',
        100: '#fce7f3',
        200: '#fbcfe8',
        300: '#f9a8d4',
        400: '#f472b6',
        500: '#ec4899',
        600: '#db2777',
        700: '#be185d',
        800: '#9d174d',
        900: '#831843',
    },
    fuchsia: {
        50: '#fdf4ff',
        100: '#fae8ff',
        200: '#f5d0fe',
        300: '#f0abfc',
        400: '#e879f9',
        500: '#d946ef',
        600: '#c026d3',
        700: '#a21caf',
        800: '#86198f',
        900: '#701a75',
    },
    purple: {
        50: '#faf5ff',
        100: '#f3e8ff',
        200: '#e9d5ff',
        300: '#d8b4fe',
        400: '#c084fc',
        500: '#a855f7',
        600: '#9333ea',
        700: '#7e22ce',
        800: '#6b21a8',
        900: '#581c87',
    },
    violet: {
        50: '#f5f3ff',
        100: '#ede9fe',
        200: '#ddd6fe',
        300: '#c4b5fd',
        400: '#a78bfa',
        500: '#8b5cf6',
        600: '#7c3aed',
        700: '#6d28d9',
        800: '#5b21b6',
        900: '#4c1d95',
    },
    indigo: {
        50: '#eef2ff',
        100: '#e0e7ff',
        200: '#c7d2fe',
        300: '#a5b4fc',
        400: '#818cf8',
        500: '#6366f1',
        600: '#4f46e5',
        700: '#4338ca',
        800: '#3730a3',
        900: '#312e81',
    },
    blue: {
        50: '#eff6ff',
        100: '#dbeafe',
        200: '#bfdbfe',
        300: '#93c5fd',
        400: '#60a5fa',
        500: '#3b82f6',
        600: '#2563eb',
        700: '#1d4ed8',
        800: '#1e40af',
        900: '#1e3a8a',
    },
    lightBlue: {
        50: '#f0f9ff',
        100: '#e0f2fe',
        200: '#bae6fd',
        300: '#7dd3fc',
        400: '#38bdf8',
        500: '#0ea5e9',
        600: '#0284c7',
        700: '#0369a1',
        800: '#075985',
        900: '#0c4a6e',
    },
    cyan: {
        50: '#ecfeff',
        100: '#cffafe',
        200: '#a5f3fc',
        300: '#67e8f9',
        400: '#22d3ee',
        500: '#06b6d4',
        600: '#0891b2',
        700: '#0e7490',
        800: '#155e75',
        900: '#164e63',
    },
    teal: {
        50: '#f0fdfa',
        100: '#ccfbf1',
        200: '#99f6e4',
        300: '#5eead4',
        400: '#2dd4bf',
        500: '#14b8a6',
        600: '#0d9488',
        700: '#0f766e',
        800: '#115e59',
        900: '#134e4a',
    },
    emerald: {
        50: '#ecfdf5',
        100: '#d1fae5',
        200: '#a7f3d0',
        300: '#6ee7b7',
        400: '#34d399',
        500: '#10b981',
        600: '#059669',
        700: '#047857',
        800: '#065f46',
        900: '#064e3b',
    },
    green: {
        50: '#f0fdf4',
        100: '#dcfce7',
        200: '#bbf7d0',
        300: '#86efac',
        400: '#4ade80',
        500: '#22c55e',
        600: '#16a34a',
        700: '#15803d',
        800: '#166534',
        900: '#14532d',
    },
    lime: {
        50: '#f7fee7',
        100: '#ecfccb',
        200: '#d9f99d',
        300: '#bef264',
        400: '#a3e635',
        500: '#84cc16',
        600: '#65a30d',
        700: '#4d7c0f',
        800: '#3f6212',
        900: '#365314',
    },
    yellow: {
        50: '#fefce8',
        100: '#fef9c3',
        200: '#fef08a',
        300: '#fde047',
        400: '#facc15',
        500: '#eab308',
        600: '#ca8a04',
        700: '#a16207',
        800: '#854d0e',
        900: '#713f12',
    },
    amber: {
        50: '#fffbeb',
        100: '#fef3c7',
        200: '#fde68a',
        300: '#fcd34d',
        400: '#fbbf24',
        500: '#f59e0b',
        600: '#d97706',
        700: '#b45309',
        800: '#92400e',
        900: '#78350f',
    },
    orange: {
        50: '#fff7ed',
        100: '#ffedd5',
        200: '#fed7aa',
        300: '#fdba74',
        400: '#fb923c',
        500: '#f97316',
        600: '#ea580c',
        700: '#c2410c',
        800: '#9a3412',
        900: '#7c2d12',
    },
    red: {
        50: '#fef2f2',
        100: '#fee2e2',
        200: '#fecaca',
        300: '#fca5a5',
        400: '#f87171',
        500: '#ef4444',
        600: '#dc2626',
        700: '#b91c1c',
        800: '#991b1b',
        900: '#7f1d1d',
    },
    warmGray: {
        50: '#fafaf9',
        100: '#f5f5f4',
        200: '#e7e5e4',
        300: '#d6d3d1',
        400: '#a8a29e',
        500: '#78716c',
        600: '#57534e',
        700: '#44403c',
        800: '#292524',
        900: '#1c1917',
    },
    trueGray: {
        50: '#fafafa',
        100: '#f5f5f5',
        200: '#e5e5e5',
        300: '#d4d4d4',
        400: '#a3a3a3',
        500: '#737373',
        600: '#525252',
        700: '#404040',
        800: '#262626',
        900: '#171717',
    },
    gray: {
        50: '#fafafa',
        100: '#f4f4f5',
        200: '#e4e4e7',
        300: '#d4d4d8',
        400: '#a1a1aa',
        500: '#71717a',
        600: '#52525b',
        700: '#3f3f46',
        800: '#27272a',
        900: '#18181b',
    },
    coolGray: {
        50: '#f9fafb',
        100: '#f3f4f6',
        200: '#e5e7eb',
        300: '#d1d5db',
        400: '#9ca3af',
        500: '#6b7280',
        600: '#4b5563',
        700: '#374151',
        800: '#1f2937',
        900: '#111827',
    },
    blueGray: {
        50: '#f8fafc',
        100: '#f1f5f9',
        200: '#e2e8f0',
        300: '#cbd5e1',
        400: '#94a3b8',
        500: '#64748b',
        600: '#475569',
        700: '#334155',
        800: '#1e293b',
        900: '#0f172a',
    },
};
var darkColors$1 = {
    black: '#000',
    white: '#181a1b',
    rose: {
        50: '#3b0004',
        100: '#430005',
        200: '#50020b',
        300: '#69020f',
        400: '#870417',
        500: '#a30924',
        600: '#b4173a',
        700: '#980e30',
        800: '#7f0e2e',
        900: '#6d0f2c',
    },
    pink: {
        50: '#340820',
        100: '#3c0725',
        200: '#4c062e',
        300: '#640739',
        400: '#840a4a',
        500: '#9c1055',
        600: '#ad1d5d',
        700: '#98134a',
        800: '#7e123e',
        900: '#691336',
    },
    fuchsia: {
        50: '#2f033a',
        100: '#330041',
        200: '#3f024e',
        300: '#550464',
        400: '#710681',
        500: '#8b0e9e',
        600: '#9a1ea9',
        700: '#82168c',
        800: '#6b1472',
        900: '#5a155e',
    },
    purple: {
        50: '#1b1d1e',
        100: '#1e2122',
        200: '#242728',
        300: '#2d3133',
        400: '#3f037c',
        500: '#510797',
        600: '#6111a9',
        700: '#651ba5',
        800: '#561a86',
        900: '#46166c',
    },
    violet: {
        50: '#1b1e1f',
        100: '#1e2122',
        200: '#242728',
        300: '#2d3133',
        400: '#220576',
        500: '#320892',
        600: '#460fa5',
        700: '#561eac',
        800: '#491a92',
        900: '#3d1777',
    },
    indigo: {
        50: '#1d1f20',
        100: '#212325',
        200: '#282b2d',
        300: '#323639',
        400: '#07127c',
        500: '#0d0f8c',
        600: '#1d169c',
        700: '#332ba0',
        800: '#2c2682',
        900: '#272567',
    },
    blue: {
        50: '#1d1f20',
        100: '#222527',
        200: '#2a2e30',
        300: '#023773',
        400: '#054391',
        500: '#0844a6',
        600: '#1043b2',
        700: '#173ead',
        800: '#18338c',
        900: '#182e6e',
    },
    lightBlue: {
        50: '#1c1f20',
        100: '#212425',
        200: '#2c3032',
        300: '#035880',
        400: '#0677a9',
        500: '#0b84ba',
        600: '#026a9f',
        700: '#025481',
        800: '#06476a',
        900: '#0a3b58',
    },
    cyan: {
        50: '#003b3e',
        100: '#02484f',
        200: '#035d67',
        300: '#067d8c',
        400: '#0e9eb4',
        500: '#0592aa',
        600: '#06748e',
        700: '#0b5d73',
        800: '#114b54',
        900: '#123e4f',
    },
    teal: {
        50: '#07362e',
        100: '#064e42',
        200: '#0a6c5e',
        300: '#128e7f',
        400: '#22a89c',
        500: '#109385',
        600: '#0a766d',
        700: '#0c5e58',
        800: '#0e4b47',
        900: '#0f3e3b',
    },
    emerald: {
        50: '#063a27',
        100: '#084a30',
        200: '#0d6245',
        300: '#168362',
        400: '#23a481',
        500: '#0d9467',
        600: '#047854',
        700: '#036046',
        800: '#054c38',
        900: '#053e2f',
    },
    green: {
        50: '#07361e',
        100: '#064425',
        200: '#0a5631',
        300: '#107645',
        400: '#1c985d',
        500: '#1b9e4b',
        600: '#12823b',
        700: '#116631',
        800: '#12512a',
        900: '#104224',
    },
    lime: {
        50: '#313f03',
        100: '#3d4f05',
        200: '#516b07',
        300: '#698c0c',
        400: '#7da715',
        500: '#6aa312',
        600: '#51820a',
        700: '#3e630c',
        800: '#324e0e',
        900: '#2b4210',
    },
    yellow: {
        50: '#3f3903',
        100: '#564f01',
        200: '#796a01',
        300: '#a18702',
        400: '#be9904',
        500: '#bb8f06',
        600: '#a26e03',
        700: '#814e06',
        800: '#6a3e0b',
        900: '#5a320e',
    },
    amber: {
        50: '#3f3200',
        100: '#544301',
        200: '#786102',
        300: '#9d7903',
        400: '#b58403',
        500: '#c37e08',
        600: '#ae5f05',
        700: '#904207',
        800: '#75330b',
        900: '#602a0c',
    },
    orange: {
        50: '#3e2200',
        100: '#4c2c00',
        200: '#653701',
        300: '#864502',
        400: '#a74d03',
        500: '#be5105',
        600: '#bb460a',
        700: '#9b340a',
        800: '#7b2a04',
        900: '#63240e',
    },
    red: {
        50: '#370404',
        100: '#430202',
        200: '#520202',
        300: '#670303',
        400: '#860707',
        500: '#9f0e0e',
        600: '#ae1c1c',
        700: '#941616',
        800: '#7a1616',
        900: '#661717',
    },
    warmGray: {
        50: '#1b1d1e',
        100: '#1e2022',
        200: '#262a2b',
        300: '#313537',
        400: '#4c5255',
        500: '#565d61',
        600: '#3e4346',
        700: '#303436',
        800: '#1d1f21',
        900: '#131516',
    },
    trueGray: {
        50: '#1b1d1e',
        100: '#1e2021',
        200: '#272a2c',
        300: '#303436',
        400: '#4c5255',
        500: '#575e61',
        600: '#3e4345',
        700: '#303436',
        800: '#1d1f20',
        900: '#111313',
    },
    gray: {
        50: '#1b1d1e',
        100: '#1e2022',
        200: '#262a2b',
        300: '#2f3335',
        400: '#4b5154',
        500: '#586064',
        600: '#414749',
        700: '#323638',
        800: '#1e2122',
        900: '#131516',
    },
    coolGray: {
        50: '#1b1d1e',
        100: '#1e2022',
        200: '#25282a',
        300: '#2f3335',
        400: '#eb5154',
        500: '#586064',
        600: '#3c444f',
        700: '#2c3441',
        800: '#19212c',
        900: '#0e131f',
    },
    blueGray: {
        50: '#1b1d1e',
        100: '#1e2022',
        200: '#242729',
        300: '#2f3335',
        400: '#3f4c5f',
        500: '#505d6f',
        600: '#394454',
        700: '#293444',
        800: '#18212f',
        900: '#0c1222',
    },
};

var variantOrder;
(function (variantOrder) {
    variantOrder[variantOrder["first"] = 0] = "first";
    variantOrder[variantOrder["last"] = 1] = "last";
    variantOrder[variantOrder["odd"] = 2] = "odd";
    variantOrder[variantOrder["even"] = 3] = "even";
    variantOrder[variantOrder["visited"] = 4] = "visited";
    variantOrder[variantOrder["checked"] = 5] = "checked";
    variantOrder[variantOrder["group-hover"] = 6] = "group-hover";
    variantOrder[variantOrder["group-focus"] = 7] = "group-focus";
    variantOrder[variantOrder["focus-within"] = 8] = "focus-within";
    variantOrder[variantOrder["hover"] = 9] = "hover";
    variantOrder[variantOrder["focus"] = 10] = "focus";
    variantOrder[variantOrder["focus-visible"] = 11] = "focus-visible";
    variantOrder[variantOrder["active"] = 12] = "active";
    variantOrder[variantOrder["disabled"] = 13] = "disabled";
})(variantOrder || (variantOrder = {}));

function toArray(v) {
    if (Array.isArray(v))
        return v;
    return [v];
}
function indent(code, tab) {
    if (tab === void 0) { tab = 2; }
    var spaces = Array(tab).fill(' ').join('');
    return code
        .split('\n')
        .map(function (line) { return spaces + line; })
        .join('\n');
}
function wrapit(code, start, end, tab, minify) {
    if (start === void 0) { start = '{'; }
    if (end === void 0) { end = '}'; }
    if (tab === void 0) { tab = 2; }
    if (minify === void 0) { minify = false; }
    if (minify)
        return "" + start + code + end;
    return start + "\n" + indent(code, tab) + "\n" + end;
}
function camelToDash(str) {
    // Use exact the same regex as Post CSS
    return str.replace(/([A-Z])/g, '-$1').replace(/^ms-/, '-ms-').toLowerCase();
}
function searchFrom(text, target, startIndex, endIndex) {
    if (startIndex === void 0) { startIndex = 0; }
    // search from partial of string
    var subText = text.substring(startIndex, endIndex);
    var relativeIndex = subText.search(target);
    return relativeIndex === -1 ? -1 : startIndex + relativeIndex;
}
function connectList(a, b, append) {
    if (append === void 0) { append = true; }
    return append ? __spreadArray(__spreadArray([], (a !== null && a !== void 0 ? a : [])), (b !== null && b !== void 0 ? b : [])) : __spreadArray(__spreadArray([], (b !== null && b !== void 0 ? b : [])), (a !== null && a !== void 0 ? a : []));
}
function deepCopy(source) {
    return Array.isArray(source)
        ? source.map(function (item) { return deepCopy(item); })
        : source instanceof Date
            ? new Date(source.getTime())
            : source && typeof source === 'object'
                ? Object.getOwnPropertyNames(source).reduce(function (o, prop) {
                    var descriptor = Object.getOwnPropertyDescriptor(source, prop);
                    if (descriptor) {
                        Object.defineProperty(o, prop, descriptor);
                        if (source && typeof source === 'object') {
                            o[prop] = deepCopy(source[prop]);
                        }
                    }
                    return o;
                }, Object.create(Object.getPrototypeOf(source)))
                : source;
}
function isTagName(name) {
    return ['a', 'abbr', 'address', 'area', 'article', 'aside', 'audio', 'b', 'base', 'bdi', 'bdo', 'blockquote', 'body', 'br', 'button', 'canvas', 'caption', 'cite', 'code', 'col', 'colgroup', 'data', 'datalist', 'dd', 'del', 'details', 'dfn', 'dialog', 'div', 'dl', 'dt', 'em', 'embd', 'fieldset', 'figcaption', 'figure', 'footer', 'form', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'head', 'header', 'hr', 'html', 'i', 'iframe', 'img', 'input', 'ins', 'kbd', 'label', 'legend', 'li', 'link', 'main', 'map', 'mark', 'meta', 'meter', 'nav', 'noscript', 'object', 'ol', 'optgroup', 'option', 'output', 'p', 'param', 'picture', 'pre', 'progress', 'q', 'rp', 'rt', 'ruby', 's', 'samp', 'script', 'section', 'select', 'small', 'source', 'span', 'strong', 'style', 'sub', 'summary', 'sup', 'svg', 'table', 'tbody', 'td', 'template', 'textarea', 'tfoot', 'th', 'thead', 'time', 'title', 'tr', 'track', 'u', 'ul', 'var', 'video', 'wbr'].includes(name);
}
function searchPropEnd(text, startIndex) {
    if (startIndex === void 0) { startIndex = 0; }
    var index = startIndex;
    var output = -1;
    var openSingleQuote = false;
    var openDoubleQuote = false;
    while (index < text.length) {
        switch (text.charAt(index)) {
            case '\'':
                if (text.charAt(index - 1) !== '\\')
                    openSingleQuote = !openSingleQuote;
                break;
            case '"':
                if (text.charAt(index - 1) !== '\\')
                    openDoubleQuote = !openDoubleQuote;
                break;
            case ';':
                if (openSingleQuote === false && openDoubleQuote === false)
                    output = index;
                break;
        }
        if (output !== -1)
            break;
        index++;
    }
    return output;
}

var Property = /** @class */ (function () {
    function Property(name, value, comment, important) {
        if (important === void 0) { important = false; }
        this.meta = { type: 'utilities', group: 'plugin', order: 99999 };
        this.name = name;
        this.value = value;
        this.comment = comment;
        this.important = important;
    }
    Property._singleParse = function (css) {
        css = css.trim();
        if (!css)
            return;
        if (css.charAt(0) === '@')
            return InlineAtRule.parse(css);
        var split = css.search(':');
        var end = searchPropEnd(css);
        if (split === -1)
            return;
        var important = false;
        var prop = css.substring(split + 1, end === -1 ? undefined : end).trim();
        if (/!important;?$/.test(prop)) {
            important = true;
            prop = prop.replace(/!important/, '').trimRight();
        }
        return new Property(css.substring(0, split).trim(), prop, undefined, important);
    };
    Property.parse = function (css) {
        if (!/;\s*$/.test(css))
            css += ';'; // Fix for the situation where the last semicolon is omitted
        var properties = [];
        var index = 0;
        var end = searchPropEnd(css, index);
        while (end !== -1) {
            var parsed = this._singleParse(css.substring(searchFrom(css, /\S/, index), end + 1));
            if (parsed)
                properties.push(parsed);
            index = end + 1;
            end = searchPropEnd(css, index);
        }
        var count = properties.length;
        if (count > 1)
            return properties;
        if (count === 1)
            return properties[0];
    };
    Property.prototype.clone = function () {
        return deepCopy(this);
    };
    Property.prototype.toStyle = function (selector) {
        return new Style(selector, this, this.important).updateMeta(this.meta);
    };
    Property.prototype.build = function (minify) {
        var _this = this;
        if (minify === void 0) { minify = false; }
        var createProperty = function (name, value) {
            if (minify) {
                return name + ":" + value + (_this.important ? '!important' : '') + ";";
            }
            else {
                var p = name + ": " + value + (_this.important ? ' !important' : '') + ";";
                return _this.comment ? p + (" /* " + _this.comment + " */") : p;
            }
        };
        if (!this.value)
            return '';
        return typeof this.name === 'string'
            ? createProperty(this.name, this.value)
            : this.name
                .map(function (i) { return createProperty(i, _this.value); })
                .join(minify ? '' : '\n');
    };
    Property.prototype.updateMeta = function (meta) {
        this.meta = meta;
        return this;
    };
    return Property;
}());
var InlineAtRule = /** @class */ (function (_super) {
    __extends(InlineAtRule, _super);
    function InlineAtRule(name, value, important) {
        if (important === void 0) { important = false; }
        var _this = _super.call(this, name, value, undefined, important) || this;
        _this.name = name;
        return _this;
    }
    InlineAtRule.parse = function (css) {
        var _a;
        var matchName = css.match(/@[^\s;{}]+/);
        if (matchName) {
            var name_1 = matchName[0].substring(1);
            var important = false;
            var expression = matchName.index !== undefined
                ? (_a = css
                    .substring(matchName.index + name_1.length + 1)
                    .match(/(?:(['"]).*?\1|[^;])*/)) === null || _a === void 0 ? void 0 : _a[0].trim()
                : undefined;
            if (expression && /!important;?$/.test(expression)) {
                important = true;
                expression = expression.replace(/!important/, '').trimRight();
            }
            return new InlineAtRule(name_1, expression === '' ? undefined : expression, important);
        }
    };
    InlineAtRule.prototype.build = function () {
        return this.value
            ? "@" + this.name + " " + this.value + (this.important ? ' !important' : '') + ";"
            : "@" + this.name + (this.important ? ' !important' : '') + ";";
    };
    return InlineAtRule;
}(Property));
var Style = /** @class */ (function () {
    function Style(selector, property, important) {
        if (important === void 0) { important = false; }
        this.meta = { type: 'components', group: 'plugin', order: 99999 };
        this.selector = selector;
        this.important = important;
        this.property = toArray(property || []);
    }
    Object.defineProperty(Style.prototype, "rule", {
        get: function () {
            var _this = this;
            var _a, _b, _c;
            var selectors = ((_a = this.selector) !== null && _a !== void 0 ? _a : '').trim().split(/\s*,\s*/g);
            this._parentSelectors && (selectors = selectors.map(function (i) { var _a; return ((_a = _this._parentSelectors) === null || _a === void 0 ? void 0 : _a.join(' ')) + " " + i; }));
            ((_b = this._wrapSelectors) !== null && _b !== void 0 ? _b : []).forEach(function (func) { return (selectors = selectors.map(function (i) { return func(i); })); });
            this._pseudoClasses && (selectors = selectors.map(function (i) { var _a; return i + (":" + ((_a = _this._pseudoClasses) === null || _a === void 0 ? void 0 : _a.join(':'))); }));
            this._pseudoElements && (selectors = selectors.map(function (i) { var _a; return i + ("::" + ((_a = _this._pseudoElements) === null || _a === void 0 ? void 0 : _a.join('::'))); }));
            this._brotherSelectors && (selectors = selectors.map(function (i) { var _a; return i + ("." + ((_a = _this._brotherSelectors) === null || _a === void 0 ? void 0 : _a.join('.'))); }));
            this._childSelectors && (selectors = selectors.map(function (i) { var _a; return i + (" " + ((_a = _this._childSelectors) === null || _a === void 0 ? void 0 : _a.join(' '))); }));
            ((_c = this._wrapRules) !== null && _c !== void 0 ? _c : []).forEach(function (func) { return (selectors = selectors.map(function (i) { return func(i); })); });
            return selectors.join(', ');
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "pseudoClasses", {
        get: function () {
            return this._pseudoClasses;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "pseudoElements", {
        get: function () {
            return this._pseudoElements;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "parentSelectors", {
        get: function () {
            return this._parentSelectors;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "childSelectors", {
        get: function () {
            return this._childSelectors;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "brotherSelectors", {
        get: function () {
            return this._brotherSelectors;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "wrapProperties", {
        get: function () {
            return this._wrapProperties;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "wrapSelectors", {
        get: function () {
            return this._wrapSelectors;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "wrapRules", {
        get: function () {
            return this._wrapRules;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Style.prototype, "simple", {
        get: function () {
            // is this style only has property and no wrap?
            return !(this.atRules || this._pseudoClasses || this._pseudoElements || this._parentSelectors || this._childSelectors || this._brotherSelectors || this._wrapProperties || this._wrapSelectors || this._wrapRules);
        },
        enumerable: false,
        configurable: true
    });
    Style.generate = function (parent, property, root) {
        if (!root)
            root = (parent === null || parent === void 0 ? void 0 : parent.startsWith('@'))
                ? new Style().atRule(parent)
                : new Style(parent);
        var output = [];
        var _loop_1 = function (key, value) {
            if (typeof value === 'string') {
                root.add(new Property(camelToDash(key), value));
            }
            else if (Array.isArray(value)) {
                value.map(function (i) { return root === null || root === void 0 ? void 0 : root.add(new Property(camelToDash(key), i)); });
            }
            else {
                var wrap = deepCopy(root);
                wrap.property = [];
                var child = void 0;
                if (key.startsWith('@')) {
                    child = wrap.atRule(key, false);
                }
                else {
                    if (wrap.selector === undefined) {
                        wrap.selector = key;
                        child = wrap;
                    }
                    else {
                        if (/^[a-z]+$/.test(key) && !isTagName(key)) {
                            wrap.wrapProperty(function (property) { return key + "-" + property; });
                            child = wrap;
                        }
                        else {
                            var _hKey_1 = function (selector, key) { return (/&/.test(key) ? key : "& " + key).replace('&', selector); };
                            wrap.wrapSelector(function (selector) {
                                return selector
                                    .trim()
                                    .split(/\s*,\s*/g)
                                    .map(function (s) {
                                    return key
                                        .split(/\s*,\s*/g)
                                        .map(function (i) { return _hKey_1(s, i); })
                                        .join(', ');
                                })
                                    .join(', ');
                            });
                            child = wrap;
                        }
                    }
                }
                output = output.concat(Style.generate(key.startsWith('@') ? undefined : key, value, child));
            }
        };
        for (var _i = 0, _a = Object.entries(property !== null && property !== void 0 ? property : {}); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], value = _b[1];
            _loop_1(key, value);
        }
        if (root.property.length > 0)
            output.unshift(root);
        return output;
    };
    Style.prototype.atRule = function (atrule, append) {
        if (append === void 0) { append = true; }
        if (!atrule)
            return this;
        if (this.atRules) {
            append ? this.atRules.push(atrule) : this.atRules.unshift(atrule);
        }
        else {
            this.atRules = [atrule];
        }
        return this;
    };
    Style.prototype.pseudoClass = function (string) {
        if (this._pseudoClasses) {
            this._pseudoClasses.push(string);
        }
        else {
            this._pseudoClasses = [string];
        }
        return this;
    };
    Style.prototype.pseudoElement = function (string) {
        if (this._pseudoElements) {
            this._pseudoElements.push(string);
        }
        else {
            this._pseudoElements = [string];
        }
        return this;
    };
    Style.prototype.brother = function (string) {
        if (this._brotherSelectors) {
            this._brotherSelectors.push(string);
        }
        else {
            this._brotherSelectors = [string];
        }
        return this;
    };
    Style.prototype.parent = function (string) {
        if (this._parentSelectors) {
            this._parentSelectors.push(string);
        }
        else {
            this._parentSelectors = [string];
        }
        return this;
    };
    Style.prototype.child = function (string) {
        if (this._childSelectors) {
            this._childSelectors.push(string);
        }
        else {
            this._childSelectors = [string];
        }
        return this;
    };
    Style.prototype.wrapProperty = function (func) {
        if (this._wrapProperties) {
            this._wrapProperties.push(func);
        }
        else {
            this._wrapProperties = [func];
        }
        return this;
    };
    Style.prototype.wrapSelector = function (func) {
        if (this._wrapSelectors) {
            this._wrapSelectors.push(func);
        }
        else {
            this._wrapSelectors = [func];
        }
        return this;
    };
    Style.prototype.wrapRule = function (func) {
        if (this._wrapRules) {
            this._wrapRules.push(func);
        }
        else {
            this._wrapRules = [func];
        }
        return this;
    };
    Style.prototype.add = function (item) {
        item = toArray(item);
        if (this.important)
            item.forEach(function (i) { return (i.important = true); });
        this.property = __spreadArray(__spreadArray([], this.property), item);
        return this;
    };
    Style.prototype.extend = function (item, onlyProperty, append) {
        if (onlyProperty === void 0) { onlyProperty = false; }
        if (append === void 0) { append = true; }
        if (!item)
            return this;
        if (item.wrapProperties) {
            var props_1 = [];
            item.property.forEach(function (p) {
                var _a;
                var pc = new Property(p.name, p.value, p.comment);
                (_a = item.wrapProperties) === null || _a === void 0 ? void 0 : _a.forEach(function (wrap) {
                    pc.name = Array.isArray(pc.name)
                        ? pc.name.map(function (i) { return wrap(i); })
                        : wrap(pc.name);
                });
                if (item.important)
                    pc.important = true;
                props_1.push(pc);
            });
            this.property = connectList(this.property, props_1, append);
        }
        else {
            if (item.important)
                item.property.forEach(function (i) { return (i.important = true); });
            this.property = connectList(this.property, item.property, append);
        }
        if (onlyProperty)
            return this;
        item.selector && (this.selector = item.selector);
        this.meta = item.meta;
        item.atRules &&
            (this.atRules = connectList(item.atRules, this.atRules, append)); // atrule is build in reverse
        item._brotherSelectors &&
            (this._brotherSelectors = connectList(this._brotherSelectors, item._brotherSelectors, append));
        item._childSelectors &&
            (this._childSelectors = connectList(this._childSelectors, item._childSelectors, append));
        item._parentSelectors &&
            (this._parentSelectors = connectList(this._parentSelectors, item._parentSelectors, append));
        item._pseudoClasses &&
            (this._pseudoClasses = connectList(this._pseudoClasses, item._pseudoClasses, append));
        item._pseudoElements &&
            (this._pseudoElements = connectList(this._pseudoElements, item._pseudoElements, append));
        item._wrapRules &&
            (this._wrapRules = connectList(this._wrapRules, item._wrapRules, append));
        item._wrapSelectors &&
            (this._wrapSelectors = connectList(this._wrapSelectors, item._wrapSelectors, append));
        return this;
    };
    Style.prototype.clean = function () {
        // remove duplicated property
        var property = [];
        var cache = [];
        this.property.forEach(function (i) {
            var inline = i.build();
            if (!cache.includes(inline)) {
                cache.push(inline);
                property.push(i);
            }
        });
        this.property = property;
        return this;
    };
    Style.prototype.flat = function () {
        var properties = [];
        this.property.forEach(function (p) {
            if (Array.isArray(p.name)) {
                p.name.forEach(function (i) {
                    properties.push(new Property(i, p.value, p.comment));
                });
            }
            else {
                properties.push(p);
            }
        });
        this.property = properties;
        return this;
    };
    Style.prototype.clone = function (selector, property) {
        var newStyle = deepCopy(this);
        if (selector)
            newStyle.selector = selector;
        if (property)
            newStyle.property = Array.isArray(property) ? property : [property];
        return newStyle;
    };
    Style.prototype.sort = function () {
        // sort property
        this.property = this.property.sort(function (a, b) {
            return ("" + a.name).substring(0, 2) > ("" + b.name).substring(0, 2) ? 1 : -1;
        });
        return this;
    };
    Style.prototype.build = function (minify, prefixer) {
        var _this = this;
        if (minify === void 0) { minify = false; }
        if (prefixer === void 0) { prefixer = true; }
        var properties = this.property;
        if (!prefixer)
            properties = properties.filter(function (p) {
                if (p.value && /-(webkit|ms|moz|o)-/.test(p.value))
                    return false;
                if (Array.isArray(p.name)) {
                    p.name = p.name.filter(function (i) { return !/^-(webkit|ms|moz|o)-/.test(i); });
                    return true;
                }
                return !/^-(webkit|ms|moz|o)-/.test(p.name);
            });
        var result = properties.map(function (p) {
            if (_this._wrapProperties) {
                var name_2 = p.name;
                _this._wrapProperties.forEach(function (w) { return (name_2 = Array.isArray(name_2) ? name_2.map(function (n) { return w(n); }) : w(name_2)); });
                return new Property(name_2, p.value, p.comment, _this.important ? true : p.important).build(minify);
            }
            return _this.important ? new Property(p.name, p.value, p.comment, true).build(minify) : p.build(minify);
        }).join(minify ? '' : '\n');
        if (!this.selector && !this.atRules)
            return result.replace(/;}/g, '}');
        if (this.selector)
            result = (minify ? this.rule.replace(/,\s/g, ',') : this.rule + ' ') + wrapit(result, undefined, undefined, undefined, result !== '' ? minify : true);
        if (this.atRules) {
            for (var _i = 0, _a = this.atRules; _i < _a.length; _i++) {
                var rule = _a[_i];
                result = minify ? "" + rule.replace(/\s/g, '') + wrapit(result, undefined, undefined, undefined, minify) : rule + " " + wrapit(result, undefined, undefined, undefined, result !== '' ? minify : true);
            }
        }
        return minify ? result.replace(/;}/g, '}') : result;
    };
    Style.prototype.updateMeta = function (meta) {
        this.meta = meta;
        return this;
    };
    return Style;
}());
/** @class */ ((function (_super) {
    __extends(GlobalStyle, _super);
    function GlobalStyle(selector, property, important) {
        return _super.call(this, selector, property, important) || this;
    }
    return GlobalStyle;
})(Style));
/** @class */ ((function (_super) {
    __extends(Keyframes, _super);
    function Keyframes(selector, property, important) {
        return _super.call(this, selector, property, important) || this;
    }
    // root param only for consist with style
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    Keyframes.generate = function (name, children, root, prefixer) {
        if (prefixer === void 0) { prefixer = true; }
        var output = [];
        for (var _i = 0, _a = Object.entries(children); _i < _a.length; _i++) {
            var _b = _a[_i], key = _b[0], value = _b[1];
            var style = new Keyframes(key).atRule("@keyframes " + name);
            var webkitStyle = new Keyframes(key).atRule("@-webkit-keyframes " + name);
            for (var _c = 0, _d = Object.entries(value); _c < _d.length; _c++) {
                var _e = _d[_c], pkey = _e[0], pvalue = _e[1];
                var prop = pkey;
                if (pkey === 'transform') {
                    prop = prefixer ? ['-webkit-transform', 'transform'] : 'transform';
                }
                else if (['animationTimingFunction', 'animation-timing-function'].includes(pkey)) {
                    prop = prefixer ? [
                        '-webkit-animation-timing-function',
                        'animation-timing-function',
                    ] : 'animation-timing-function';
                }
                style.add(new Property(prop, pvalue));
                webkitStyle.add(new Property(prop, pvalue));
            }
            output.push(style);
            if (prefixer)
                output.push(webkitStyle);
        }
        return output;
    };
    return Keyframes;
})(Style));
/** @class */ ((function (_super) {
    __extends(Container, _super);
    function Container(selector, property, important) {
        return _super.call(this, selector, property, important) || this;
    }
    return Container;
})(Style));

function combineObject(a, b) {
    var output = __assign({}, a);
    for (var _i = 0, _a = Object.entries(b); _i < _a.length; _i++) {
        var _b = _a[_i], key_of_b = _b[0], value_of_b = _b[1];
        if (key_of_b in a) {
            var value_of_a = a[key_of_b];
            if (value_of_a !== value_of_b) {
                if (value_of_b !== null &&
                    value_of_b.constructor !== Object) {
                    output[key_of_b] = [value_of_a, value_of_b];
                }
                else if (value_of_a !== null &&
                    value_of_a.constructor === Object) {
                    output[key_of_b] = combineObject(value_of_a, value_of_b);
                }
                else {
                    output[key_of_b] = [
                        value_of_a,
                        combineObject(value_of_a, value_of_b),
                    ];
                }
            }
        }
        else {
            output[key_of_b] = value_of_b;
        }
    }
    return output;
}

var lightColors = {
    transparent: 'transparent',
    current: 'currentColor',
    black: colors.black,
    white: colors.white,
    gray: colors.coolGray,
    red: colors.red,
    yellow: colors.amber,
    green: colors.emerald,
    blue: colors.blue,
    indigo: colors.indigo,
    purple: colors.violet,
    pink: colors.pink,
    rose: colors.rose,
    fuchsia: colors.fuchsia,
    violet: colors.violet,
    cyan: colors.cyan,
    teal: colors.teal,
    emerald: colors.emerald,
    lime: colors.lime,
    amber: colors.amber,
    orange: colors.orange,
    'light-blue': colors.lightBlue,
    'warm-gray': colors.warmGray,
    'true-gray': colors.trueGray,
    'cool-gray': colors.coolGray,
    'blue-gray': colors.blueGray,
};
var darkColors = {
    transparent: 'transparent',
    current: 'currentColor',
    black: darkColors$1.black,
    white: darkColors$1.white,
    gray: darkColors$1.coolGray,
    red: darkColors$1.red,
    yellow: darkColors$1.amber,
    green: darkColors$1.emerald,
    blue: darkColors$1.blue,
    indigo: darkColors$1.indigo,
    purple: darkColors$1.violet,
    pink: darkColors$1.pink,
    rose: darkColors$1.rose,
    fuchsia: darkColors$1.fuchsia,
    violet: darkColors$1.violet,
    cyan: darkColors$1.cyan,
    teal: darkColors$1.teal,
    emerald: darkColors$1.emerald,
    lime: darkColors$1.lime,
    amber: darkColors$1.amber,
    orange: darkColors$1.orange,
    'light-blue': darkColors$1.lightBlue,
    'warm-gray': darkColors$1.warmGray,
    'true-gray': darkColors$1.trueGray,
    'cool-gray': darkColors$1.coolGray,
    'blue-gray': darkColors$1.blueGray,
};
combineObject(lightColors, darkColors);

var round = function (num) {
    return num
        .toFixed(7)
        .replace(/(\.[0-9]+?)0+$/, '$1')
        .replace(/\.0$/, '');
};
var rem = function (px) { return round(px / 16) + "rem"; };
var em = function (px, base) { return round(px / base) + "em"; };
var styles = function (theme) { return (__assign({ DEFAULT: {
        css: [
            {
                color: theme('colors.gray.700', colors.gray[700]),
                maxWidth: '65ch',
                '[class~="lead"]': {
                    color: theme('colors.gray.600', colors.gray[600]),
                },
                a: {
                    color: theme('colors.gray.900', colors.gray[900]),
                    textDecoration: 'underline',
                    fontWeight: '500',
                },
                strong: {
                    color: theme('colors.gray.900', colors.gray[900]),
                    fontWeight: '600',
                },
                'ol[type="A"]': {
                    '--list-counter-style': 'upper-alpha',
                },
                'ol[type="a"]': {
                    '--list-counter-style': 'lower-alpha',
                },
                'ol[type="A s"]': {
                    '--list-counter-style': 'upper-alpha',
                },
                'ol[type="a s"]': {
                    '--list-counter-style': 'lower-alpha',
                },
                'ol[type="I"]': {
                    '--list-counter-style': 'upper-roman',
                },
                'ol[type="i"]': {
                    '--list-counter-style': 'lower-roman',
                },
                'ol[type="I s"]': {
                    '--list-counter-style': 'upper-roman',
                },
                'ol[type="i s"]': {
                    '--list-counter-style': 'lower-roman',
                },
                'ol[type="1"]': {
                    '--list-counter-style': 'decimal',
                },
                'ol > li': {
                    position: 'relative',
                },
                'ol > li::before': {
                    content: 'counter(list-item, var(--list-counter-style, decimal)) "."',
                    position: 'absolute',
                    fontWeight: '400',
                    color: theme('colors.gray.500', colors.gray[500]),
                },
                'ul > li': {
                    position: 'relative',
                },
                'ul > li::before': {
                    content: '""',
                    position: 'absolute',
                    backgroundColor: theme('colors.gray.300', colors.gray[300]),
                    borderRadius: '50%',
                },
                hr: {
                    borderColor: theme('colors.gray.200', colors.gray[200]),
                    borderTopWidth: 1,
                },
                blockquote: {
                    fontWeight: '500',
                    fontStyle: 'italic',
                    color: theme('colors.gray.900', colors.gray[900]),
                    borderLeftWidth: '0.25rem',
                    borderLeftColor: theme('colors.gray.200', colors.gray[200]),
                    quotes: '"\\201C""\\201D""\\2018""\\2019"',
                },
                'blockquote p:first-of-type::before': {
                    content: 'open-quote',
                },
                'blockquote p:last-of-type::after': {
                    content: 'close-quote',
                },
                h1: {
                    color: theme('colors.gray.900', colors.gray[900]),
                    fontWeight: '800',
                },
                h2: {
                    color: theme('colors.gray.900', colors.gray[900]),
                    fontWeight: '700',
                },
                h3: {
                    color: theme('colors.gray.900', colors.gray[900]),
                    fontWeight: '600',
                },
                h4: {
                    color: theme('colors.gray.900', colors.gray[900]),
                    fontWeight: '600',
                },
                'figure figcaption': {
                    color: theme('colors.gray.500', colors.gray[500]),
                },
                code: {
                    color: theme('colors.gray.900', colors.gray[900]),
                    fontWeight: '600',
                },
                'code::before': {
                    content: '"`"',
                },
                'code::after': {
                    content: '"`"',
                },
                'a code': {
                    color: theme('colors.gray.900', colors.gray[900]),
                },
                pre: {
                    color: theme('colors.gray.200', colors.gray[200]),
                    backgroundColor: theme('colors.gray.800', colors.gray[800]),
                    overflowX: 'auto',
                },
                'pre code': {
                    backgroundColor: 'transparent',
                    borderWidth: '0',
                    borderRadius: '0',
                    padding: '0',
                    fontWeight: '400',
                    color: 'inherit',
                    fontSize: 'inherit',
                    fontFamily: 'inherit',
                    lineHeight: 'inherit',
                },
                'pre code::before': {
                    content: 'none',
                },
                'pre code::after': {
                    content: 'none',
                },
                table: {
                    width: '100%',
                    tableLayout: 'auto',
                    textAlign: 'left',
                    marginTop: em(32, 16),
                    marginBottom: em(32, 16),
                },
                thead: {
                    color: theme('colors.gray.900', colors.gray[900]),
                    fontWeight: '600',
                    borderBottomWidth: '1px',
                    borderBottomColor: theme('colors.gray.300', colors.gray[300]),
                },
                'thead th': {
                    verticalAlign: 'bottom',
                },
                'tbody tr': {
                    borderBottomWidth: '1px',
                    borderBottomColor: theme('colors.gray.200', colors.gray[200]),
                },
                'tbody tr:last-child': {
                    borderBottomWidth: '0',
                },
                'tbody td': {
                    verticalAlign: 'top',
                },
            },
            {
                fontSize: rem(16),
                lineHeight: round(28 / 16),
                p: {
                    marginTop: em(20, 16),
                    marginBottom: em(20, 16),
                },
                '[class~="lead"]': {
                    fontSize: em(20, 16),
                    lineHeight: round(32 / 20),
                    marginTop: em(24, 20),
                    marginBottom: em(24, 20),
                },
                blockquote: {
                    marginTop: em(32, 20),
                    marginBottom: em(32, 20),
                    paddingLeft: em(20, 20),
                },
                h1: {
                    fontSize: em(36, 16),
                    marginTop: '0',
                    marginBottom: em(32, 36),
                    lineHeight: round(40 / 36),
                },
                h2: {
                    fontSize: em(24, 16),
                    marginTop: em(48, 24),
                    marginBottom: em(24, 24),
                    lineHeight: round(32 / 24),
                },
                h3: {
                    fontSize: em(20, 16),
                    marginTop: em(32, 20),
                    marginBottom: em(12, 20),
                    lineHeight: round(32 / 20),
                },
                h4: {
                    marginTop: em(24, 16),
                    marginBottom: em(8, 16),
                    lineHeight: round(24 / 16),
                },
                img: {
                    marginTop: em(32, 16),
                    marginBottom: em(32, 16),
                },
                video: {
                    marginTop: em(32, 16),
                    marginBottom: em(32, 16),
                },
                figure: {
                    marginTop: em(32, 16),
                    marginBottom: em(32, 16),
                },
                'figure > *': {
                    marginTop: '0',
                    marginBottom: '0',
                },
                'figure figcaption': {
                    fontSize: em(14, 16),
                    lineHeight: round(20 / 14),
                    marginTop: em(12, 14),
                },
                code: {
                    fontSize: em(14, 16),
                },
                'h2 code': {
                    fontSize: em(21, 24),
                },
                'h3 code': {
                    fontSize: em(18, 20),
                },
                pre: {
                    fontSize: em(14, 16),
                    lineHeight: round(24 / 14),
                    marginTop: em(24, 14),
                    marginBottom: em(24, 14),
                    borderRadius: rem(6),
                    paddingTop: em(12, 14),
                    paddingRight: em(16, 14),
                    paddingBottom: em(12, 14),
                    paddingLeft: em(16, 14),
                },
                ol: {
                    marginTop: em(20, 16),
                    marginBottom: em(20, 16),
                    'list-style-type': 'none',
                },
                ul: {
                    marginTop: em(20, 16),
                    marginBottom: em(20, 16),
                    'list-style-type': 'none',
                },
                li: {
                    marginTop: em(8, 16),
                    marginBottom: em(8, 16),
                },
                'ol > li': {
                    paddingLeft: em(28, 16),
                },
                'ol > li::before': {
                    left: '0',
                },
                'ul > li': {
                    paddingLeft: em(28, 16),
                },
                'ul > li::before': {
                    width: em(6, 16),
                    height: em(6, 16),
                    top: "calc(" + em(28 / 2, 16) + " - " + em(3, 16) + ")",
                    left: em(4, 16),
                },
                '> ul > li p': {
                    marginTop: em(12, 16),
                    marginBottom: em(12, 16),
                },
                '> ul > li > *:first-child': {
                    marginTop: em(20, 16),
                },
                '> ul > li > *:last-child': {
                    marginBottom: em(20, 16),
                },
                '> ol > li > *:first-child': {
                    marginTop: em(20, 16),
                },
                '> ol > li > *:last-child': {
                    marginBottom: em(20, 16),
                },
                'ul ul, ul ol, ol ul, ol ol': {
                    marginTop: em(12, 16),
                    marginBottom: em(12, 16),
                },
                hr: {
                    marginTop: em(48, 16),
                    marginBottom: em(48, 16),
                },
                'hr + *': {
                    marginTop: '0',
                },
                'h2 + *': {
                    marginTop: '0',
                },
                'h3 + *': {
                    marginTop: '0',
                },
                'h4 + *': {
                    marginTop: '0',
                },
                table: {
                    fontSize: em(14, 16),
                    lineHeight: round(24 / 14),
                },
                'thead th': {
                    paddingRight: em(8, 14),
                    paddingBottom: em(8, 14),
                    paddingLeft: em(8, 14),
                },
                'thead th:first-child': {
                    paddingLeft: '0',
                },
                'thead th:last-child': {
                    paddingRight: '0',
                },
                'tbody td': {
                    paddingTop: em(8, 14),
                    paddingRight: em(8, 14),
                    paddingBottom: em(8, 14),
                    paddingLeft: em(8, 14),
                },
                'tbody td:first-child': {
                    paddingLeft: '0',
                },
                'tbody td:last-child': {
                    paddingRight: '0',
                },
            },
            {
                '> :first-child': {
                    marginTop: '0',
                },
                '> :last-child': {
                    marginBottom: '0',
                },
            },
        ],
    }, sm: {
        css: [
            {
                fontSize: rem(14),
                lineHeight: round(24 / 14),
                p: {
                    marginTop: em(16, 14),
                    marginBottom: em(16, 14),
                },
                '[class~="lead"]': {
                    fontSize: em(18, 14),
                    lineHeight: round(28 / 18),
                    marginTop: em(16, 18),
                    marginBottom: em(16, 18),
                },
                blockquote: {
                    marginTop: em(24, 18),
                    marginBottom: em(24, 18),
                    paddingLeft: em(20, 18),
                },
                h1: {
                    fontSize: em(30, 14),
                    marginTop: '0',
                    marginBottom: em(24, 30),
                    lineHeight: round(36 / 30),
                },
                h2: {
                    fontSize: em(20, 14),
                    marginTop: em(32, 20),
                    marginBottom: em(16, 20),
                    lineHeight: round(28 / 20),
                },
                h3: {
                    fontSize: em(18, 14),
                    marginTop: em(28, 18),
                    marginBottom: em(8, 18),
                    lineHeight: round(28 / 18),
                },
                h4: {
                    marginTop: em(20, 14),
                    marginBottom: em(8, 14),
                    lineHeight: round(20 / 14),
                },
                img: {
                    marginTop: em(24, 14),
                    marginBottom: em(24, 14),
                },
                video: {
                    marginTop: em(24, 14),
                    marginBottom: em(24, 14),
                },
                figure: {
                    marginTop: em(24, 14),
                    marginBottom: em(24, 14),
                },
                'figure > *': {
                    marginTop: '0',
                    marginBottom: '0',
                },
                'figure figcaption': {
                    fontSize: em(12, 14),
                    lineHeight: round(16 / 12),
                    marginTop: em(8, 12),
                },
                code: {
                    fontSize: em(12, 14),
                },
                'h2 code': {
                    fontSize: em(18, 20),
                },
                'h3 code': {
                    fontSize: em(16, 18),
                },
                pre: {
                    fontSize: em(12, 14),
                    lineHeight: round(20 / 12),
                    marginTop: em(20, 12),
                    marginBottom: em(20, 12),
                    borderRadius: rem(4),
                    paddingTop: em(8, 12),
                    paddingRight: em(12, 12),
                    paddingBottom: em(8, 12),
                    paddingLeft: em(12, 12),
                },
                ol: {
                    marginTop: em(16, 14),
                    marginBottom: em(16, 14),
                },
                ul: {
                    marginTop: em(16, 14),
                    marginBottom: em(16, 14),
                },
                li: {
                    marginTop: em(4, 14),
                    marginBottom: em(4, 14),
                },
                'ol > li': {
                    paddingLeft: em(22, 14),
                },
                'ol > li::before': {
                    left: '0',
                },
                'ul > li': {
                    paddingLeft: em(22, 14),
                },
                'ul > li::before': {
                    height: em(5, 14),
                    width: em(5, 14),
                    top: "calc(" + em(24 / 2, 14) + " - " + em(2.5, 14) + ")",
                    left: em(3, 14),
                },
                '> ul > li p': {
                    marginTop: em(8, 14),
                    marginBottom: em(8, 14),
                },
                '> ul > li > *:first-child': {
                    marginTop: em(16, 14),
                },
                '> ul > li > *:last-child': {
                    marginBottom: em(16, 14),
                },
                '> ol > li > *:first-child': {
                    marginTop: em(16, 14),
                },
                '> ol > li > *:last-child': {
                    marginBottom: em(16, 14),
                },
                'ul ul, ul ol, ol ul, ol ol': {
                    marginTop: em(8, 14),
                    marginBottom: em(8, 14),
                },
                hr: {
                    marginTop: em(40, 14),
                    marginBottom: em(40, 14),
                },
                'hr + *': {
                    marginTop: '0',
                },
                'h2 + *': {
                    marginTop: '0',
                },
                'h3 + *': {
                    marginTop: '0',
                },
                'h4 + *': {
                    marginTop: '0',
                },
                table: {
                    fontSize: em(12, 14),
                    lineHeight: round(18 / 12),
                },
                'thead th': {
                    paddingRight: em(12, 12),
                    paddingBottom: em(8, 12),
                    paddingLeft: em(12, 12),
                },
                'thead th:first-child': {
                    paddingLeft: '0',
                },
                'thead th:last-child': {
                    paddingRight: '0',
                },
                'tbody td': {
                    paddingTop: em(8, 12),
                    paddingRight: em(12, 12),
                    paddingBottom: em(8, 12),
                    paddingLeft: em(12, 12),
                },
                'tbody td:first-child': {
                    paddingLeft: '0',
                },
                'tbody td:last-child': {
                    paddingRight: '0',
                },
            },
            {
                '> :first-child': {
                    marginTop: '0',
                },
                '> :last-child': {
                    marginBottom: '0',
                },
            },
        ],
    }, lg: {
        css: [
            {
                fontSize: rem(18),
                lineHeight: round(32 / 18),
                p: {
                    marginTop: em(24, 18),
                    marginBottom: em(24, 18),
                },
                '[class~="lead"]': {
                    fontSize: em(22, 18),
                    lineHeight: round(32 / 22),
                    marginTop: em(24, 22),
                    marginBottom: em(24, 22),
                },
                blockquote: {
                    marginTop: em(40, 24),
                    marginBottom: em(40, 24),
                    paddingLeft: em(24, 24),
                },
                h1: {
                    fontSize: em(48, 18),
                    marginTop: '0',
                    marginBottom: em(40, 48),
                    lineHeight: round(48 / 48),
                },
                h2: {
                    fontSize: em(30, 18),
                    marginTop: em(56, 30),
                    marginBottom: em(32, 30),
                    lineHeight: round(40 / 30),
                },
                h3: {
                    fontSize: em(24, 18),
                    marginTop: em(40, 24),
                    marginBottom: em(16, 24),
                    lineHeight: round(36 / 24),
                },
                h4: {
                    marginTop: em(32, 18),
                    marginBottom: em(8, 18),
                    lineHeight: round(28 / 18),
                },
                img: {
                    marginTop: em(32, 18),
                    marginBottom: em(32, 18),
                },
                video: {
                    marginTop: em(32, 18),
                    marginBottom: em(32, 18),
                },
                figure: {
                    marginTop: em(32, 18),
                    marginBottom: em(32, 18),
                },
                'figure > *': {
                    marginTop: '0',
                    marginBottom: '0',
                },
                'figure figcaption': {
                    fontSize: em(16, 18),
                    lineHeight: round(24 / 16),
                    marginTop: em(16, 16),
                },
                code: {
                    fontSize: em(16, 18),
                },
                'h2 code': {
                    fontSize: em(26, 30),
                },
                'h3 code': {
                    fontSize: em(21, 24),
                },
                pre: {
                    fontSize: em(16, 18),
                    lineHeight: round(28 / 16),
                    marginTop: em(32, 16),
                    marginBottom: em(32, 16),
                    borderRadius: rem(6),
                    paddingTop: em(16, 16),
                    paddingRight: em(24, 16),
                    paddingBottom: em(16, 16),
                    paddingLeft: em(24, 16),
                },
                ol: {
                    marginTop: em(24, 18),
                    marginBottom: em(24, 18),
                },
                ul: {
                    marginTop: em(24, 18),
                    marginBottom: em(24, 18),
                },
                li: {
                    marginTop: em(12, 18),
                    marginBottom: em(12, 18),
                },
                'ol > li': {
                    paddingLeft: em(30, 18),
                },
                'ol > li::before': {
                    left: '0',
                },
                'ul > li': {
                    paddingLeft: em(30, 18),
                },
                'ul > li::before': {
                    width: em(6, 18),
                    height: em(6, 18),
                    top: "calc(" + em(32 / 2, 18) + " - " + em(3, 18) + ")",
                    left: em(4, 18),
                },
                '> ul > li p': {
                    marginTop: em(16, 18),
                    marginBottom: em(16, 18),
                },
                '> ul > li > *:first-child': {
                    marginTop: em(24, 18),
                },
                '> ul > li > *:last-child': {
                    marginBottom: em(24, 18),
                },
                '> ol > li > *:first-child': {
                    marginTop: em(24, 18),
                },
                '> ol > li > *:last-child': {
                    marginBottom: em(24, 18),
                },
                'ul ul, ul ol, ol ul, ol ol': {
                    marginTop: em(16, 18),
                    marginBottom: em(16, 18),
                },
                hr: {
                    marginTop: em(56, 18),
                    marginBottom: em(56, 18),
                },
                'hr + *': {
                    marginTop: '0',
                },
                'h2 + *': {
                    marginTop: '0',
                },
                'h3 + *': {
                    marginTop: '0',
                },
                'h4 + *': {
                    marginTop: '0',
                },
                table: {
                    fontSize: em(16, 18),
                    lineHeight: round(24 / 16),
                },
                'thead th': {
                    paddingRight: em(12, 16),
                    paddingBottom: em(12, 16),
                    paddingLeft: em(12, 16),
                },
                'thead th:first-child': {
                    paddingLeft: '0',
                },
                'thead th:last-child': {
                    paddingRight: '0',
                },
                'tbody td': {
                    paddingTop: em(12, 16),
                    paddingRight: em(12, 16),
                    paddingBottom: em(12, 16),
                    paddingLeft: em(12, 16),
                },
                'tbody td:first-child': {
                    paddingLeft: '0',
                },
                'tbody td:last-child': {
                    paddingRight: '0',
                },
            },
            {
                '> :first-child': {
                    marginTop: '0',
                },
                '> :last-child': {
                    marginBottom: '0',
                },
            },
        ],
    }, xl: {
        css: [
            {
                fontSize: rem(20),
                lineHeight: round(36 / 20),
                p: {
                    marginTop: em(24, 20),
                    marginBottom: em(24, 20),
                },
                '[class~="lead"]': {
                    fontSize: em(24, 20),
                    lineHeight: round(36 / 24),
                    marginTop: em(24, 24),
                    marginBottom: em(24, 24),
                },
                blockquote: {
                    marginTop: em(48, 30),
                    marginBottom: em(48, 30),
                    paddingLeft: em(32, 30),
                },
                h1: {
                    fontSize: em(56, 20),
                    marginTop: '0',
                    marginBottom: em(48, 56),
                    lineHeight: round(56 / 56),
                },
                h2: {
                    fontSize: em(36, 20),
                    marginTop: em(56, 36),
                    marginBottom: em(32, 36),
                    lineHeight: round(40 / 36),
                },
                h3: {
                    fontSize: em(30, 20),
                    marginTop: em(48, 30),
                    marginBottom: em(20, 30),
                    lineHeight: round(40 / 30),
                },
                h4: {
                    marginTop: em(36, 20),
                    marginBottom: em(12, 20),
                    lineHeight: round(32 / 20),
                },
                img: {
                    marginTop: em(40, 20),
                    marginBottom: em(40, 20),
                },
                video: {
                    marginTop: em(40, 20),
                    marginBottom: em(40, 20),
                },
                figure: {
                    marginTop: em(40, 20),
                    marginBottom: em(40, 20),
                },
                'figure > *': {
                    marginTop: '0',
                    marginBottom: '0',
                },
                'figure figcaption': {
                    fontSize: em(18, 20),
                    lineHeight: round(28 / 18),
                    marginTop: em(18, 18),
                },
                code: {
                    fontSize: em(18, 20),
                },
                'h2 code': {
                    fontSize: em(31, 36),
                },
                'h3 code': {
                    fontSize: em(27, 30),
                },
                pre: {
                    fontSize: em(18, 20),
                    lineHeight: round(32 / 18),
                    marginTop: em(36, 18),
                    marginBottom: em(36, 18),
                    borderRadius: rem(8),
                    paddingTop: em(20, 18),
                    paddingRight: em(24, 18),
                    paddingBottom: em(20, 18),
                    paddingLeft: em(24, 18),
                },
                ol: {
                    marginTop: em(24, 20),
                    marginBottom: em(24, 20),
                },
                ul: {
                    marginTop: em(24, 20),
                    marginBottom: em(24, 20),
                },
                li: {
                    marginTop: em(12, 20),
                    marginBottom: em(12, 20),
                },
                'ol > li': {
                    paddingLeft: em(36, 20),
                },
                'ol > li::before': {
                    left: '0',
                },
                'ul > li': {
                    paddingLeft: em(36, 20),
                },
                'ul > li::before': {
                    width: em(7, 20),
                    height: em(7, 20),
                    top: "calc(" + em(36 / 2, 20) + " - " + em(3.5, 20) + ")",
                    left: em(5, 20),
                },
                '> ul > li p': {
                    marginTop: em(16, 20),
                    marginBottom: em(16, 20),
                },
                '> ul > li > *:first-child': {
                    marginTop: em(24, 20),
                },
                '> ul > li > *:last-child': {
                    marginBottom: em(24, 20),
                },
                '> ol > li > *:first-child': {
                    marginTop: em(24, 20),
                },
                '> ol > li > *:last-child': {
                    marginBottom: em(24, 20),
                },
                'ul ul, ul ol, ol ul, ol ol': {
                    marginTop: em(16, 20),
                    marginBottom: em(16, 20),
                },
                hr: {
                    marginTop: em(56, 20),
                    marginBottom: em(56, 20),
                },
                'hr + *': {
                    marginTop: '0',
                },
                'h2 + *': {
                    marginTop: '0',
                },
                'h3 + *': {
                    marginTop: '0',
                },
                'h4 + *': {
                    marginTop: '0',
                },
                table: {
                    fontSize: em(18, 20),
                    lineHeight: round(28 / 18),
                },
                'thead th': {
                    paddingRight: em(12, 18),
                    paddingBottom: em(16, 18),
                    paddingLeft: em(12, 18),
                },
                'thead th:first-child': {
                    paddingLeft: '0',
                },
                'thead th:last-child': {
                    paddingRight: '0',
                },
                'tbody td': {
                    paddingTop: em(16, 18),
                    paddingRight: em(12, 18),
                    paddingBottom: em(16, 18),
                    paddingLeft: em(12, 18),
                },
                'tbody td:first-child': {
                    paddingLeft: '0',
                },
                'tbody td:last-child': {
                    paddingRight: '0',
                },
            },
            {
                '> :first-child': {
                    marginTop: '0',
                },
                '> :last-child': {
                    marginBottom: '0',
                },
            },
        ],
    }, '2xl': {
        css: [
            {
                fontSize: rem(24),
                lineHeight: round(40 / 24),
                p: {
                    marginTop: em(32, 24),
                    marginBottom: em(32, 24),
                },
                '[class~="lead"]': {
                    fontSize: em(30, 24),
                    lineHeight: round(44 / 30),
                    marginTop: em(32, 30),
                    marginBottom: em(32, 30),
                },
                blockquote: {
                    marginTop: em(64, 36),
                    marginBottom: em(64, 36),
                    paddingLeft: em(40, 36),
                },
                h1: {
                    fontSize: em(64, 24),
                    marginTop: '0',
                    marginBottom: em(56, 64),
                    lineHeight: round(64 / 64),
                },
                h2: {
                    fontSize: em(48, 24),
                    marginTop: em(72, 48),
                    marginBottom: em(40, 48),
                    lineHeight: round(52 / 48),
                },
                h3: {
                    fontSize: em(36, 24),
                    marginTop: em(56, 36),
                    marginBottom: em(24, 36),
                    lineHeight: round(44 / 36),
                },
                h4: {
                    marginTop: em(40, 24),
                    marginBottom: em(16, 24),
                    lineHeight: round(36 / 24),
                },
                img: {
                    marginTop: em(48, 24),
                    marginBottom: em(48, 24),
                },
                video: {
                    marginTop: em(48, 24),
                    marginBottom: em(48, 24),
                },
                figure: {
                    marginTop: em(48, 24),
                    marginBottom: em(48, 24),
                },
                'figure > *': {
                    marginTop: '0',
                    marginBottom: '0',
                },
                'figure figcaption': {
                    fontSize: em(20, 24),
                    lineHeight: round(32 / 20),
                    marginTop: em(20, 20),
                },
                code: {
                    fontSize: em(20, 24),
                },
                'h2 code': {
                    fontSize: em(42, 48),
                },
                'h3 code': {
                    fontSize: em(32, 36),
                },
                pre: {
                    fontSize: em(20, 24),
                    lineHeight: round(36 / 20),
                    marginTop: em(40, 20),
                    marginBottom: em(40, 20),
                    borderRadius: rem(8),
                    paddingTop: em(24, 20),
                    paddingRight: em(32, 20),
                    paddingBottom: em(24, 20),
                    paddingLeft: em(32, 20),
                },
                ol: {
                    marginTop: em(32, 24),
                    marginBottom: em(32, 24),
                },
                ul: {
                    marginTop: em(32, 24),
                    marginBottom: em(32, 24),
                },
                li: {
                    marginTop: em(12, 24),
                    marginBottom: em(12, 24),
                },
                'ol > li': {
                    paddingLeft: em(40, 24),
                },
                'ol > li::before': {
                    left: '0',
                },
                'ul > li': {
                    paddingLeft: em(40, 24),
                },
                'ul > li::before': {
                    width: em(8, 24),
                    height: em(8, 24),
                    top: "calc(" + em(40 / 2, 24) + " - " + em(4, 24) + ")",
                    left: em(6, 24),
                },
                '> ul > li p': {
                    marginTop: em(20, 24),
                    marginBottom: em(20, 24),
                },
                '> ul > li > *:first-child': {
                    marginTop: em(32, 24),
                },
                '> ul > li > *:last-child': {
                    marginBottom: em(32, 24),
                },
                '> ol > li > *:first-child': {
                    marginTop: em(32, 24),
                },
                '> ol > li > *:last-child': {
                    marginBottom: em(32, 24),
                },
                'ul ul, ul ol, ol ul, ol ol': {
                    marginTop: em(16, 24),
                    marginBottom: em(16, 24),
                },
                hr: {
                    marginTop: em(72, 24),
                    marginBottom: em(72, 24),
                },
                'hr + *': {
                    marginTop: '0',
                },
                'h2 + *': {
                    marginTop: '0',
                },
                'h3 + *': {
                    marginTop: '0',
                },
                'h4 + *': {
                    marginTop: '0',
                },
                table: {
                    fontSize: em(20, 24),
                    lineHeight: round(28 / 20),
                },
                'thead th': {
                    paddingRight: em(12, 20),
                    paddingBottom: em(16, 20),
                    paddingLeft: em(12, 20),
                },
                'thead th:first-child': {
                    paddingLeft: '0',
                },
                'thead th:last-child': {
                    paddingRight: '0',
                },
                'tbody td': {
                    paddingTop: em(16, 20),
                    paddingRight: em(12, 20),
                    paddingBottom: em(16, 20),
                    paddingLeft: em(12, 20),
                },
                'tbody td:first-child': {
                    paddingLeft: '0',
                },
                'tbody td:last-child': {
                    paddingRight: '0',
                },
            },
            {
                '> :first-child': {
                    marginTop: '0',
                },
                '> :last-child': {
                    marginBottom: '0',
                },
            },
        ],
    } }, Object.entries(colors).reduce(function (reduced, _a) {
    var _b;
    var color = _a[0], values = _a[1];
    if (!isUsableColor(color, values)) {
        return reduced;
    }
    return __assign(__assign({}, reduced), (_b = {}, _b[color] = {
        css: [
            {
                a: { color: values[600] },
                'a code': { color: values[600] },
            },
        ],
    }, _b));
}, {}))); };

function combineConfig(a, b) {
    var output = __assign({}, a);
    for (var _i = 0, _a = Object.entries(b); _i < _a.length; _i++) {
        var _b = _a[_i], key_of_b = _b[0], value_of_b = _b[1];
        if (key_of_b in a) {
            var value_of_a = a[key_of_b];
            if (value_of_a !== value_of_b) {
                if (value_of_b !== null && value_of_b.constructor !== Object) {
                    if (Array.isArray(value_of_a) && Array.isArray(value_of_b)) {
                        output[key_of_b] = __spreadArray(__spreadArray([], value_of_a), value_of_b);
                    }
                    else {
                        output[key_of_b] = value_of_b;
                    }
                }
                else if (value_of_a !== null && value_of_a.constructor === Object) {
                    output[key_of_b] = combineConfig(value_of_a, value_of_b);
                }
                else if (Array.isArray(value_of_a)) {
                    output[key_of_b] = __spreadArray(__spreadArray([], value_of_a), Array.isArray(value_of_b) ? value_of_b : [value_of_b]);
                }
                else {
                    output[key_of_b] = __assign({ DEFAULT: value_of_a }, value_of_b);
                }
            }
        }
        else {
            output[key_of_b] = value_of_b;
        }
    }
    return output;
}

var computed = {
// Reserved for future "magic properties", for example:
// bulletColor: (color) => ({ 'ul > li::before': { backgroundColor: color } }),
};
function configToCss(config) {
    if (config === void 0) { config = {}; }
    return __spreadArray(__spreadArray([], Object.keys(config)
        .filter(function (key) { return computed[key]; })
        .map(function (key) { return computed[key](config[key]); })), castArray(config.css || {})).reduce(function (previous, current) { return combineConfig(previous, current); }, {});
}
var index = createPlugin.withOptions(function (_a) {
    var _b = _a === void 0 ? {} : _a, modifiers = _b.modifiers, _c = _b.className, className = _c === void 0 ? 'prose' : _c;
    return function (_a) {
        var addComponents = _a.addComponents, theme = _a.theme, variants = _a.variants;
        var DEFAULT_MODIFIERS = __spreadArray([
            'DEFAULT',
            'sm',
            'lg',
            'xl',
            '2xl'
        ], Object.entries(theme('colors'))
            .filter(function (_a) {
            var color = _a[0], values = _a[1];
            return isUsableColor(color, values);
        })
            .map(function (_a) {
            var color = _a[0];
            return color;
        }));
        modifiers = modifiers === undefined ? DEFAULT_MODIFIERS : modifiers;
        var config = theme('typography');
        var all = uniq(__spreadArray(__spreadArray([
            'DEFAULT'
        ], modifiers), Object.keys(config).filter(function (modifier) { return !DEFAULT_MODIFIERS.includes(modifier); })));
        addComponents(all.map(function (modifier) {
            var _a;
            return (_a = {},
                _a[modifier === 'DEFAULT'
                    ? "." + className
                    : "." + className + "-" + modifier] = configToCss(config[modifier]),
                _a);
        }), variants('typography'));
    };
}, function () { return ({
    theme: { typography: styles },
    variants: { typography: ['responsive'] },
}); });

module.exports = index;
